src/accumulate.jl

```diff
@@ -2,104 +2,157 @@
+        cache[thread] = val
+    end
+    sync_warp()
+    if lane > 4 
+        val = op(cache[thread - 4], val)
+        cache[thread] = val
+    end
+    sync_warp()
+    if lane > 8 
+        val = op(cache[thread - 8], val)
+        cache[thread] = val
+    end
+    sync_warp()
+    if lane > 16 
+        val = op(cache[thread - 16], val)
+        cache[thread] = val
+    end
+    sync_warp()
+    return val
+end
+end
+
+
+function partial_scan!(op::Function, output::CuDeviceArray{T}, input::CuDeviceArray,
+                      aggregates::Union{Nothing, CuDeviceArray{T}}, Rdim, Rpre, 
+                      Rpost, Rother, init, ::Val{inclusive}=Val(true), 
+                      ::Val{shuffle}=Val(true)) where {T, inclusive, shuffle}
     threads = blockDim().x
     thread = threadIdx().x
     block = blockIdx().x
+    wid, lane = fldmod1(thread, warpsize())
+    exclusive = !inclusive
 
-    temp = @cuDynamicSharedMem(T, (2*threads,))
+    # cache: storage for non shuffle kernels
+    partial_sums = @cuDynamicSharedMem(T, 32)
+    cache = @cuDynamicSharedMem(T, threads*!shuffle)
 
     # iterate the main dimension using threads and the first block dimension
     i = (blockIdx().x-1) * blockDim().x + threadIdx().x
     # iterate the other dimensions using the remaining block dimensions
     j = (blockIdx().z-1) * gridDim().y + blockIdx().y
 
-    if j > length(Rother)
-        return
-    end
+    j > length(Rother) && return
 
     @inbounds begin
         I = Rother[j]
         Ipre = Rpre[I[1]]
         Ipost = Rpost[I[2]]
-    end
 
-    # load input into shared memory (apply `op` to have the correct type)
-    @inbounds temp[thread] = if i <= length(Rdim)
-        op(neutral, input[Ipre, i, Ipost])
-    else
-        op(neutral, neutral)
-    end
+        # convert input val to correct type
+        # take element from previous index in exclusive scan
+        exclusive && (i = i - 1) 
+        value = if i <= length(Rdim) && i != 0
+            convert(T, input[Ipre, i, Ipost])
+        else
+            convert(T, input[Ipre, 1, Ipost])
+        end
+        exclusive && (i = i + 1) 
 
-    # build sum in place up the tree
-    offset = 1
-    d = threads>>1
-    while d > 0
-        sync_threads()
-        @inbounds if thread <= d
-            ai = offset * (2*thread-1)
-            bi = offset * (2*thread)
-            temp[bi] = op(temp[ai], temp[bi])
+        # Apply init to the first element
+        if inclusive
+            i == 1 && init !== nothing && (value = op(init, value))
+        else
+            i == 1 && (value = init)
         end
-        offset *= 2
-        d >>= 1
-    end
 
-    # clear the last element
-    @inbounds if thread == 1
-        temp[threads] = neutral
-    end
+        if shuffle
+            value = scan_warp(op, value, lane)
+        else
+            cache[thread] = value
+            value = scan_warp(op, value, lane, thread, cache)
+            sync_threads()
```

Comment:
Why is this `sync_threads` needed here? You are already doing a `sync_warp` inside `scan_warp`.

Reply:
The comment I posted at the time of initially ([above](https://github.com/JuliaGPU/CUDA.jl/pull/287#discussion_r453214094)) is the same as this so I'll address this here.

Exactly, we should not need a `sync_threads` here, a `sync_warp` should suffice. But, it's not working out, the results are inconsistent without the `sync_threads`.

Reply:
> Don't you need this sync_threads because warp 1 is going to access shared memory written to by all the warps?
No, if you look at the next few lines
```julia
                sync_threads()
        end

        lane == warpsize() && (partial_sums[wid] = value)

        sync_threads()
```
each warp writes to a location exclusive to id and then we call sync_threads().  `sync_threads` after writing to `partial_sums` is necessary but before that shouldn't be.

Reply:
Right your comment https://github.com/JuliaGPU/CUDA.jl/pull/287/files#r453214094  is associated in the unified diff with the `sync_threads` below, and that one is necessary since we write to shared memory. 

The one above seems indeed unnecessary. So we are indeed on the same page.

Reply:
I used cuda-memcheck's synccheck option and it cannot detect any sync issues. [link](https://gist.github.com/Ellipse0934/eadf4a9aa78a3816880f075111d6e0d0) However, I am easily able to reproduce the invalid result.